src/transformers.jl

```diff
@@ -28,20 +28,73 @@ Non-mutating version of [`transform!`](@ref).
 function transform end
 
 """
-    Transforms.apply!(data::T, ::Transform; kwargs...) -> T
+    apply(data::T, ::Transform; kwargs...) -> T
+
+Applies the [`Transform`](@ref) to the data. New transforms should usually only extend
+`_apply` which this method delegates to.
+
+Where possible, this should be extended for new data types `T`.
+"""
+function apply end
+
+"""
+    apply!(data::T, ::Transform; kwargs...) -> T
+
+Applies the [`Transform`](@ref) mutating the input `data`. New transforms should usually
+only extend `_apply!` which this method delegates to.
 
-Applies the [`Transform`](@ref) mutating the input `data`.
 Where possible, this should be extended for new data types `T`.
 """
 function apply! end
 
+
 """
-    Transforms.apply(data::T, ::Transform; kwargs...) -> T
+    apply(A::AbstractArray, ::Transform; dims=:, inds=:, kwargs...)
+
+Applies the [`Transform`](@ref) to each element of `A`.
+
+Optionally specify the `dims` to apply the [`Transform`](@ref) along certain dimensions
+and `inds` will be the indices to apply the Transform to along the `dims` specified.
 
-Non-mutating version of [`apply!`](@ref), which it delegates to by default.
-Does not need to be extended unless a mutating [`Transform`](@ref) is not possible.
+If `dims` === : (all dimensions), then `inds` will be the global indices of the array,
+instead of being relative to a certain dimension.
 """
-function apply end
+function apply(A::AbstractArray, t::Transform; dims=:, inds=:, kwargs...)
+    if dims === Colon()
+        if inds === Colon()
+            return _apply(A, t; kwargs...)
+        else
+            if A isa KeyedArray
+                # KeyedArrays don't support indexing into them via an array of indices
```

Comment:
what about `A[:][inds]`?

```julia
julia> A = KeyedArray(rand(3, 3); node=["a", "b", "c"], time=[1, 2, 3]);

julia> inds=[2, 3];

julia> A[inds]
ERROR: MethodError: no method matching keys_getindex(::Tuple{Array{Int64,1}}, ::Tuple{})
Closest candidates are:
  keys_getindex(::Tuple{}, ::Tuple{}) at /Users/glenn/.julia/packages/AxisKeys/iJ0Vf/src/struct.jl:122
  keys_getindex(::Tuple, ::Tuple{Integer,Vararg{Any,N} where N}) at /Users/glenn/.julia/packages/AxisKeys/iJ0Vf/src/struct.jl:95
  keys_getindex(::Tuple, ::Tuple{Base.Slice,Vararg{Any,N} where N}) at /Users/glenn/.julia/packages/AxisKeys/iJ0Vf/src/struct.jl:99
  ...
Stacktrace:
 [1] keys_getindex at /Users/glenn/.julia/packages/AxisKeys/iJ0Vf/src/struct.jl:109 [inlined]
 [2] getindex(::KeyedArray{Float64,2,NamedDimsArray{(:node, :time),Float64,2,Array{Float64,2}},Tuple{Array{String,1},Array{Int64,1}}}, ::Array{Int64,1}) at /Users/glenn/.julia/packages/AxisKeys/iJ0Vf/src/struct.jl:80
 [3] top-level scope at REPL[34]:1

julia> A[:][inds]
2-element NamedDimsArray(::Array{Float64,1}, (:node,)):
↓ node  0.8601795204696292
        0.10124610540366952

julia> M[:][inds]
2-element Array{Int64,1}:
 4
 7
```

maybe this is less memory efficient? but `A[:]` would be allocating a new array anyway.

Also means we can get rid of the dependency on `AxisKeys`.

Reply:
This doesn't work for vectors

Reply:
>This doesn't work for vectors

what MWE did you use?

```julia
julia> v = [1, 2, 3, 4, 5];

julia> inds=[2, 3];

julia> v[:][inds]
2-element Array{Int64,1}:
 2
 3
```

Reply:
Nvm, whatever example it was it seems to be working now